test/Main.hs

```diff
@@ -52,7 +53,22 @@ testImplies :: Bit -> Bit -> Expectation
 testImplies a b
   | a == I = implies a b `shouldBe` b
   | otherwise = implies a b `shouldBe` I
-  
+
+testHalfAdder :: Bit -> Bit -> Expectation
+testHalfAdder a b
+  | a == I && b == I = halfAdder a b `shouldBe` (O, I)
+  | a == O && b == O = halfAdder a b `shouldBe` (O, O)
+  | otherwise = halfAdder a b `shouldBe` (I, O)
+
+testFullAdder :: Bit -> Bit -> Bit -> Expectation
+testFullAdder a b c
+  | a == I && b == I && c == I = fullAdder a b c `shouldBe` (I, I)
+  | a == I && b == I && c == O = fullAdder a b c `shouldBe` (O, I)
+  | a == I && b == O && c == I = fullAdder a b c `shouldBe` (O, I)
+  | a == O && b == I && c == I = fullAdder a b c `shouldBe` (O, I)
+  | a == O && b == O && c == O = fullAdder a b c `shouldBe` (O, O)
+  | otherwise = fullAdder a b c `shouldBe` (I, O)
```

Comment:
These tests are basically just truth tables. We're not looking at any properties. You could even write:

```
testHalfAdder :: Expectation
testHalfAdder = do
    halfAdder I I `shouldBe` (O, I)
    halfAdder O O `shouldBe` (O, O)
    ...
```

Let's find a way to make these a little more abstract than unit tests.

Reply:
BTW, a better way to write truth tables would be something like this:

```haskell
testHalfAdder :: Bit -> Bit -> Expectation
testHalfAdder I I = halfAdder I I `shouldBe` (O, I)
testHalfAdder O O = halfAdder I I `shouldBe` (O, O)
...
```

But again, that's not what we want in a property test.

Reply:
A property test should be a fairly short description of how we expect the inputs to influence the output. Test code is still code, written for humans to understand and maintain.

If I read these truth tables, it is not immediately obvious to me that `| a == O && b == I && c == I = fullAdder a b c `shouldBe` (O, I)` is correct. I should be able to read a test and immediately tell if it is correct.

Reply:
I was under the impression that a truth table essentially is a property, thanks for correcting me.
What about those properties:
`comHalfAdder :: Bit -> Bit -> Expectation
comHalfAdder a b = halfAdder a b `shouldBe` halfAdder b a

idHalfAdder :: Bit -> Bit -> Expectation
idHalfAdder a b
  | b == O = halfAdder a b `shouldBe` (a, b)
  | otherwise = halfAdder a b `shouldNotBe` (a, b)

comFullAdder :: Bit -> Bit -> Bit -> Expectation
comFullAdder a b c = fullAdder a b c `shouldBe` fullAdder c b a

carryFullAdder :: Bit -> Bit -> Bit -> Expectation
carryFullAdder a b c
  | c == O = fullAdder a b c `shouldBe` halfAdder a b
  | otherwise = fullAdder a b c `shouldNotBe` halfAdder a b
`?